tests/TestCase/Network/Email/EmailTest.php

```diff
@@ -2592,6 +2596,74 @@ public function testZeroOnlyLinesNotBeingEmptied()
     }
 
     /**
+     * testJsonSerialize()
+     *
+     * @return void
+     */
+    public function testJsonSerialize()
+    {
+        $xmlstr = <<<XML
+<?xml version='1.0' standalone='yes'?>
+<framework>
+    <name>CakePHP</name>
+    <url>http://cakephp.org</url>
+</framework>
+XML;
+
+        $this->CakeEmail->reset()
+            ->to(['cakephp@cakephp.org' => 'CakePHP'])
+            ->from('noreply@cakephp.org')
+            ->replyTo('cakephp@cakephp.org')
+            ->cc(['mark@cakephp.org', 'juan@cakephp.org' => 'Juan Basso'])
+            ->bcc('phpnut@cakephp.org')
+            ->subject('Test Serialize')
+            ->template('default', 'test')
+            ->messageId('<uuid@server.com>')
+            ->domain('foo.bar')
+            ->viewVars([
+                'users' => TableRegistry::get('Users')->get(1, ['fields' => ['id', 'username']]),
+                'file' => fopen(TEST_APP . 'config' . DS . 'empty.ini', 'r'),
```

Comment:
Digged a little further into this and I was wrong assuming it would just handle it.

JSON does not support binary data and so it would need to be `base64_encode()`'d. Here is a possible solution:

```php
// when iterating in jsonSerialize()
            // Store file contents instead of resource.
            if (is_resource($item) && get_resource_type($item) == 'stream') {
                $item = base64_encode(stream_get_contents($item));
                $array['_encodedViewVars'][] = $key;
            }

// when cloning with createFromArray()
        if (!empty($this->_viewVars) && !empty($encoded)) {
            array_walk_recursive($this->_viewVars, function (&$item, $key) use ($encoded) {
                if (in_array($key, $encoded)) {
                    $item = base64_decode($item);
                }
            });
        }
```

I know, it's not the cleanest - and thus why haven't committed it but it works. 

I tried checking also in #php on IRC but they confirm there is no way to check for binary with UTF.